plugins/org.eclipse.n4js/src/org/eclipse/n4js/typesbuilder/N4JSTypesBuilder.xtend

```diff
@@ -193,6 +242,84 @@ public class N4JSTypesBuilder {
 		}
 	}
 
+	def private void relinkTypes(Script script, TModule target, boolean preLinkingPhase) {
+		var topLevelTypesIdx = 0;
+		var variableIndex = 0;
+		for (n : script.eAllContents.toIterable) {
+			switch n {
+				TypeDefiningElement: {
+					topLevelTypesIdx = n.relinkType(target, preLinkingPhase, topLevelTypesIdx);
+				}
+				ExportedVariableStatement: {
+					variableIndex = n.relinkType(target, preLinkingPhase, variableIndex)
+				}
+			}
+		}
+	}
+
+	def protected dispatch int relinkType(TypeDefiningElement other, TModule target, boolean preLinkingPhase, int idx) {
+		throw new IllegalArgumentException("unknown subclass of TypeDefiningElement: "+other?.eClass.name);
+	}
+
+	def protected dispatch int relinkType(NamespaceImportSpecifier nsImpSpec, TModule target, boolean preLinkingPhase, int idx) {
+		// already handled up-front in #buildNamespacesTypesFromModuleImports()
+		return idx;
+	}
+
+	def protected dispatch int relinkType(N4ClassDeclaration n4Class, TModule target, boolean preLinkingPhase, int idx) {
+		if (n4Class.relinkTClass(target, preLinkingPhase, idx)) {
+			return idx + 1;
+		}
+		return idx;
+	}
+
+	def protected dispatch int relinkType(N4ClassExpression n4Class, TModule target, boolean preLinkingPhase, int idx) {
+		n4Class.createTClass(target, preLinkingPhase)
+		// do not increment the index
+		return idx
+	}
+
+	def protected dispatch int relinkType(N4InterfaceDeclaration n4Interface, TModule target, boolean preLinkingPhase, int idx) {
+		if (n4Interface.relinkTInterface(target, preLinkingPhase, idx)) {
+			return idx+1;
+		}
+		return idx;
+	}
+
+	def protected dispatch int relinkType(N4EnumDeclaration n4Enum, TModule target, boolean preLinkingPhase, int idx) {
+		if (n4Enum.relinkTEnum(target, preLinkingPhase, idx)) {
+			return idx + 1;
+		}
+		return idx;
+	}
+
+	def protected dispatch int relinkType(ObjectLiteral objectLiteral, TModule target, boolean preLinkingPhase, int idx) {
+		objectLiteral.createObjectLiteral(target, preLinkingPhase)
```

Comment:
@szarnekow Are you sure this works or is this WIP?

`#createObjectLiteral()` simply creates a new type and adds that to the `TModule#internalTypes`. Since we assume the TModule has been loaded from the index, `#internalTypes` will be empty, but for some of the types we create here, there will already be types in `#exposedInternalTypes` that should ideally be relinked.

Also, those new internal types will never be moved to `TModule#exposedInternalTypes`, or have I overlooked something?

(all of the above applies to handling of function expressions in line 315)

Reply:
I'd think this works since assignability is checked on the structure of the types and not their identity. Thus the exposedInternalTypes will contain the 'same' types as the internal types which should not make a difference. Was there a bug that proved this assumption wrong?

Reply:
Thanks for the quick reply!
True, in those two cases there will be no identity check, so the duplication should not be an issue.
But the functionality to move some to `#exposedInternalTypes` still seems to be required. Will have a closer look at this tomorrow.